client-device-connection-infinispan/src/main/java/org/eclipse/hono/deviceconnection/infinispan/client/DeviceConnectionInfoCache.java

```diff
@@ -0,0 +1,67 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.deviceconnection.infinispan.client;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A repository for keeping connection information about devices.
+ *
+ */
+public interface DeviceConnectionInfoCache {
+
+    /**
+     * Sets the gateway that last acted on behalf of a device.
+     * <p>
+     * If a device connects directly instead of through a gateway, the device identifier itself is to be used as value
+     * for the <em>gatewayId</em> parameter.
+     *
+     * @param tenant The tenant that the device belongs to.
+     * @param deviceId The device identifier.
+     * @param gatewayId The gateway identifier. This may be the same as the device identifier if the device is
+     *                  (currently) not connected via a gateway but directly to a protocol adapter.
+     * @param context The currently active OpenTracing span or {@code null} if no span is currently active.
+     *            Implementing classes should use this as the parent for any span they create for tracing
+     *            the execution of this operation.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future will be succeeded if the device connection information has been updated.
+     *         Otherwise the future will be failed with a {@link org.eclipse.hono.client.ServiceInvocationException}.
+     * @throws NullPointerException if tenant, device id or gateway id are {@code null}.
+     */
+    Future<Void> setLastKnownGatewayForDevice(String tenant, String deviceId, String gatewayId, SpanContext context);
+
+    /**
+     * Gets the gateway that last acted on behalf of a device.
+     * <p>
+     * If no last known gateway has been set for the given device yet, a failed future with status
+     * <em>404</em> is returned.
+     *
+     * @param tenant The tenant that the device belongs to.
+     * @param deviceId The device identifier.
+     * @param context The currently active OpenTracing span or {@code null} if no span is currently active.
+     *            Implementing classes should use this as the parent for any span they create for tracing
+     *            the execution of this operation.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future will be succeeded with a JSON object containing the currently mapped gateway ID
+     *         in the <em>gateway-id</em> property, if device connection information has been found for
+     *         the given device.
+     *         Otherwise the future will be failed with a {@link org.eclipse.hono.client.ServiceInvocationException}.
+     * @throws NullPointerException if tenant or device id are {@code null}.
+     */
+    Future<JsonObject> getLastKnownGatewayForDevice(String tenant, String deviceId, SpanContext context);
```

Comment:
Wouldn't it make more sense to return `Optional<JsonObject>`, and indicate a missing entry by returning an empty optional? Creating an exception always has the overhead of filling in the stack trace. And "not finding" anything in cache isn't necessarily a failure either.

Reply:
Then we would not be able to pass on any (potential) exception that occurred while accessing the (remote) cache. On the other hand, there is not much that client code of this method could do with this exception anyway, right? We already pass in a SpanContext so an implementation can already properly log the problem. That would probably also account for the *setLastKnownGateway* method as well, and we should change its return value to `void` accordingly, shouldn't we?